python/KMS_KEYS_TO_NOT_DELETE/KMS_KEYS_TO_NOT_DELETE.py

```diff
@@ -0,0 +1,455 @@
+         Then: Return NOT_APPLICABLE
+
+    	Scenario 3:
+    	Given: At least 1 CMK is present in the AWS Account
+      	  And: Rule parameter kmsKeyIds are configured and valid
+      	  And: The KMS key in the parameter is not in list of kms keys
+     	 Then: Return NOT_APPLICABLE
+
+	Scenario 4:
+	Given: At least 1 CMK is present in the AWS Account
+	  And: Rule parameter kmsKeyIds are not configured
+	  And: The KMS Key is not scheduled for deletion
+	 Then: Return COMPLIANT
+
+	Scenario 5:
+	Given: At least 1 CMK is present in the AWS Account
+	  And: Rule parameter kmsKeyIds are not configured
+	  And: The KMS Key in the account is scheduled for deletion
+	 Then: Return NON_COMPLIANT
+
+	Scenario 6:
+	Given: At least 1 CMK is present in the AWS Account
+	  And: Rule parameter kmsKeyIds are configured and valid
+	  And: The CMK is one of the keys in the parameter
+	  And: The CMK is not scheduled for deletion
+	 Then: Return COMPLIANT
+
+	Scenario 7:
+	Given: At least 1 CMK is present in the AWS Account
+	  And: Rule parameter kmsKeyIds are configured and valid
+	  And: The CMK is one of the keys in the parameter
+	  And: The CMK is scheduled for deletion
+	 Then: Return NON_COMPLIANT
+'''
+
+import json
+import sys
+import datetime
+import re
+import boto3
+import botocore
+
+try:
+    import liblogging
+except ImportError:
+    pass
+
+##############
+# Parameters #
+##############
+
+# Define the default resource to report to Config Rules
+DEFAULT_RESOURCE_TYPE = 'AWS::KMS::Key'
+
+# Set to True to get the lambda to assume the Role attached on the Config Service (useful for cross-account).
+ASSUME_ROLE_MODE = False
+
+# Other parameters (no change needed)
+CONFIG_ROLE_TIMEOUT_SECONDS = 900
+
+#############
+# Main Code #
+#############
+
+def evaluate_compliance(event, configuration_item, valid_rule_parameters):
+    evaluations = []
+    kms_client = get_client('kms', event)
+    all_kms_key_list = get_all_kms_keys(kms_client)
+
+    if not all_kms_key_list:
+        return None
+
+    if 'kmsKeyIds' in valid_rule_parameters:
+        for key_id in valid_rule_parameters['kmsKeyIds']:
+            if key_id in all_kms_key_list:
+                kms_key_details = kms_client.describe_key(KeyId=key_id)
+                if kms_key_details['KeyMetadata']['Origin'] == 'AWS_KMS' and kms_key_details['KeyMetadata']['KeyManager'] == 'CUSTOMER':
+                    if kms_key_details['KeyMetadata']['KeyState'] == 'PendingDeletion':
+                        evaluations.append(build_evaluation(key_id, 'NON_COMPLIANT', event, annotation='The KMS Key is scheduled for deletion.'))
+                        continue
+                    evaluations.append(build_evaluation(key_id, 'COMPLIANT', event))
+            if not key_id in all_kms_key_list:
+                evaluations.append(build_evaluation(key_id, 'NOT_APPLICABLE', event, 'AWS::::Account', annotation='The given kmsKeyId does not exist. Please verify the kmsKeyId and try again.'))

```

Comment:
Keep the resource type by default. Just remove this `AWS::::Account`.

Reply:
We always report NOT_APPLICABLE on Resource Account right. So in this case it will be "AWS::KMS::Key"? because we are reporting on the parameter?

Reply:
It is because the resource id is the kms id.